src/shogun/features/DenseFeatures.cpp

```diff
@@ -865,6 +865,33 @@ SGVector<ST> CDenseFeatures<ST>::dot(const SGVector<ST>& other) const
 	return linalg::matrix_prod(get_feature_matrix(), other, false);
 }
 
+template <typename ST>
+void CDenseFeatures<ST>::dot(
+	SGVector<float64_t> output, IntRange feat_range,
+	const SGVector<float64_t> vec, const SGVector<float64_t> alphas,
+	float64_t b) const
+{
+	if (m_subset_stack->has_subsets() || !feature_matrix.matrix)
+		return CDotFeatures::dot(output, feat_range, vec, alphas, b);
+
+	ASSERT(feat_range.m_begin >= 0)
+	ASSERT(feat_range.m_end <= get_num_vectors())
+
+	auto sliced_matrix =
+		feature_matrix.submatrix(feat_range.m_begin, feat_range.m_end);
+
+	if constexpr (std::is_same<ST, float64_t>::value)
+		linalg::matrix_prod(sliced_matrix, vec, output, true);
+	else
+		linalg::matrix_prod(
+			sliced_matrix.template as<float64_t>(), vec, output, true);
+
+	if (alphas.size() != 0)
+		linalg::element_prod(output, alphas, output);
+	if (b != 0)
+		linalg::add_scalar(output, b);
+}
```

Comment:
really like this now. Doing the batch thing if possible, otherwise the normal thing...quite cute!